server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -503,37 +448,120 @@ private void innerExecute(IndexRequest indexRequest, Pipeline pipeline, Consumer
         }
     }
 
-    private void innerUpdatePipelines(ClusterState previousState, ClusterState state) {
+    @Override
+    public void applyClusterState(final ClusterChangedEvent event) {
+        ClusterState state = event.state();
         if (state.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
             return;
         }
 
-        IngestMetadata ingestMetadata = state.getMetaData().custom(IngestMetadata.TYPE);
-        IngestMetadata previousIngestMetadata = previousState.getMetaData().custom(IngestMetadata.TYPE);
-        if (Objects.equals(ingestMetadata, previousIngestMetadata)) {
+        IngestMetadata newIngestMetadata = state.getMetaData().custom(IngestMetadata.TYPE);
+        if (newIngestMetadata == null) {
             return;
         }
 
-        Map<String, Pipeline> pipelines = new HashMap<>();
+        try {
+            innerUpdatePipelines(newIngestMetadata);
+        } catch (ElasticsearchParseException e) {
+            logger.warn("failed to update ingest pipelines", e);
+        }
+    }
+
+    void innerUpdatePipelines(IngestMetadata newIngestMetadata) {
+        Map<String, PipelineHolder> pipelines = new HashMap<>(this.pipelines);
```

Comment:
If I understand correctly on cluster state changes..

the old implementation - compared ingest meta data and if nothing changed returned immediately

this implementation - compares each pipeline, updates the ones as needed, updates the "global" pipelines that will be used for the next batch. 

The difference is probably trivial, but there is some work in copying maps done for each cluster state change. Is it possible to short circuit this method like in the old implementation to return immediately if nothing changed. It adds a bit more work when the pipeline changes, but 99% of the time it isn't changing and can save some work by short circuiting this.

Reply:
I probably wasn't clear ... but it is possible to re-introduce part of the old behavior to short circuit and return immediately if the ingest meta data did not change ?

Reply:
So in the latest commit I did add the `removed` and `updated` variables in order to avoid the map copying in case that there are no changes (via `Maps#copyOf(...)`). I missed the creation of the first map in this method... I will change this method to do as little work as possible in case there are no changes to pipelines (which is the most likely scenario).

But I think we do need to keep comparing each pipeline configuration for changes? I don't see a way to shortcircuit that without adding back the old check. Also `IngestMetadata#equal(...)` does also compare also all pipeline configurations, so in terms of comparing work nothing really changed here.

Reply:
I pushed: 9fab208381dac0ccb5fbb56174987dc41c9e2bd4